### PR TITLE
All interface objects now share the same hasInstance

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
r? @nox

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/9633)

#9281

### CODE REVIEW

File: components/script/dom/bindings/interface.rs [hunk 1]
```diff
@@ -273,22 +278,25 @@ pub unsafe fn has_instance(
     }
     let mut value = RootedObject::new(cx, value.to_object());
 
-    // Steps 2-3 only concern callback interface objects.
+    let class = JS_GetClass(*prototype.ptr);
+    if let Ok(object) = get_dom_class(UncheckedUnwrapObject(class, 0)) {
```
Comment (GuillaumeGomez) [inline components/script/dom/bindings/interface.rs]: How do you get `NonCallbackInterfaceObjectClass` instance here?

Comment (nox) [inline components/script/dom/bindings/interface.rs]: `&*(&class as *const _ as *const NonCallbackInterfaceObjectClass)`

**Comment by GuillaumeGomez** [comment]
Done.

**Comment by GuillaumeGomez** [comment]
Done.

**Comment by nox** [comment]
Great work, just a few corrections needed.

-S-awaiting-review +S-needs-code-changes 

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[components/script/dom/bindings/codegen/CodegenRust.py, line 1898 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAWZtvgGcEGNs491OAd:-KAWZtvgGcEGNs491OAe:114487736) ([raw file](https://github.com/servo/servo/blob/06965d810e0791daa7e536b99325f9627cea9d15/components/script/dom/bindings/codegen/CodegenRust.py#L1898)):*
`depth`

---

*[components/script/dom/bindings/codegen/CodegenRust.py, line 1902 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAWZq-vD9aaV9WALz3l:-KAWZq-vD9aaV9WALz3m:2133986040) ([raw file](https://github.com/servo/servo/blob/06965d810e0791daa7e536b99325f9627cea9d15/components/script/dom/bindings/codegen/CodegenRust.py#L1902)):*
Everything related to `hasInstance` should go away from here.

---

*[components/script/dom/bindings/codegen/CodegenRust.py, line 4740 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAWZlL2corpkbKqt-xs:-KAWZlL2corpkbKqt-xt:-1533381231) ([raw file](https://github.com/servo/servo/blob/06965d810e0791daa7e536b99325f9627cea9d15/components/script/dom/bindings/codegen/CodegenRust.py#L4740)):*
With this patch, this whole class should go away.

---

*[components/script/dom/bindings/codegen/CodegenRust.py, line 5363 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAWZyhGeq3bCgy1-cSw:-KAWZyhGeq3bCgy1-cSx:634278233) ([raw file](https://github.com/servo/servo/blob/06965d810e0791daa7e536b99325f9627cea9d15/components/script/dom/bindings/codegen/CodegenRust.py#L5363)):*
No need to import this here, nor `has_instance`.

---

*[components/script/dom/bindings/interface.rs, line 282 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAVHCG9_YT41fdaZOfw-r1-282:-KAWYtbpapio13G_RYwn:-1878695854) ([raw file](https://github.com/servo/servo/blob/0f562cf7210682130a83a5a23f5df0affc02aea0/components/script/dom/bindings/interface.rs#L282)):*
Call these two variables `js_class` and `object_class`.

---

*[components/script/dom/bindings/interface.rs, line 88 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAWYZEcezBxwWWYaBLD:-KAWYZEcezBxwWWYaBLE:331933746) ([raw file](https://github.com/servo/servo/blob/06965d810e0791daa7e536b99325f9627cea9d15/components/script/dom/bindings/interface.rs#L88)):*
This type isn't useful anymore and can go away.

---

*[components/script/dom/bindings/interface.rs, line 107 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAWYA6HJxH6z-4QTTdd:-KAWYA6HJxH6z-4QTTde:-2029990575) ([raw file](https://github.com/servo/servo/blob/06965d810e0791daa7e536b99325f9627cea9d15/components/script/dom/bindings/interface.rs#L107)):*
"The prototype ID of that interface, used in the hasInstance hook."

---

*[components/script/dom/bindings/interface.rs, line 109 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAWYFu7AyhBtWjnotGY:-KAWYFu7AyhBtWjnotGZ:-1676736539) ([raw file](https://github.com/servo/servo/blob/06965d810e0791daa7e536b99325f9627cea9d15/components/script/dom/bindings/interface.rs#L109)):*
"The prototype depth of that interface, used in the hasInstance hook."

---

*[components/script/dom/bindings/interface.rs, line 110 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAWYIiCjzcI222FoQqg:-KAWYIiCjzcI222FoQqh:1668576381) ([raw file](https://github.com/servo/servo/blob/06965d810e0791daa7e536b99325f9627cea9d15/components/script/dom/bindings/interface.rs#L110)):*
`proto_depth: u16`

---

*[components/script/dom/bindings/interface.rs, line 119 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAWYOxpsVnqqhObivgB:-KAWYOxpsVnqqhObivgC:-93176463) ([raw file](https://github.com/servo/servo/blob/06965d810e0791daa7e536b99325f9627cea9d15/components/script/dom/bindings/interface.rs#L119)):*
There shouldn't be a variable `has_instance` parameter anymore, instead `class.hasInstance` should be set to `has_instance_hook`.

---

*[components/script/dom/bindings/interface.rs, line 122 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAWYNc--CTFCIW3HADu:-KAWYNc--CTFCIW3HADv:786872239) ([raw file](https://github.com/servo/servo/blob/06965d810e0791daa7e536b99325f9627cea9d15/components/script/dom/bindings/interface.rs#L122)):*
`proto_depth`

---

*[components/script/dom/bindings/interface.rs, line 126 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAWZLAA_pO31-6SXXgY:-KAWZLAA_pO31-6SXXgZ:-854977326) ([raw file](https://github.com/servo/servo/blob/06965d810e0791daa7e536b99325f9627cea9d15/components/script/dom/bindings/interface.rs#L126)):*
Nit: could you put these two fields after the `class` one? It weirds me out.

---

*[components/script/dom/bindings/interface.rs, line 271 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAWYbgsaAFWZMyAdQPE:-KAWYbgsaAFWZMyAdQPF:1262594666) ([raw file](https://github.com/servo/servo/blob/06965d810e0791daa7e536b99325f9627cea9d15/components/script/dom/bindings/interface.rs#L271)):*
`has_instance_hook`, and it doesn't need to be public anymore.
Also nit: please indent the arguments like I did for the other functions in that module.

---

*[components/script/dom/bindings/interface.rs, line 286 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAWZa3pyr6Y7Va90rd7:-KAWZa3pyr6Y7Va90rd8:-1195761796) ([raw file](https://github.com/servo/servo/blob/06965d810e0791daa7e536b99325f9627cea9d15/components/script/dom/bindings/interface.rs#L286)):*
No need to be public anymore.

---

*[components/script/dom/bindings/interface.rs, line 302 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAWZEiWMoJCc3HtEjj9:-KAWZEiWMoJCc3HtEjjA:1797633941) ([raw file](https://github.com/servo/servo/blob/06965d810e0791daa7e536b99325f9627cea9d15/components/script/dom/bindings/interface.rs#L302)):*
`[object_class.proto_depth as usize]`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9633#-:-KAW_5U6XJyEvL7pAG1p:647906406)*

**Comment by nox** [comment]
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9633#-:-KAWb-6JohJ-RpKvSF4Z:1458266033)*

**Comment by GuillaumeGomez** [comment]
Re-updated.

**Comment by nox** [comment]
A few nits, and the actual reason why your patch is broken was because of one reference too many in the cast from `JSClass` to `NonCallbackInterfaceObjectClass`.


-S-awaiting-review -S-tests-failed +S-needs-code-changes 

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[components/script/dom/bindings/interface.rs, line 112 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAj8vo6-nnJECIPXX1o:-KAj8vo6-nnJECIPXX1p:-473168703) ([raw file](https://github.com/servo/servo/blob/857cf8773e1cf1474228930c76d7e7b2159ee930/components/script/dom/bindings/interface.rs#L112)):*
`u16`

---

*[components/script/dom/bindings/interface.rs, line 260 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAj8bpFD8MgamJOeYmX:-KAj8bpFD8MgamJOeYmY:-1836633590) ([raw file](https://github.com/servo/servo/blob/857cf8773e1cf1474228930c76d7e7b2159ee930/components/script/dom/bindings/interface.rs#L260)):*
`/// Hook for instanceof on interface objects.`

---

*[components/script/dom/bindings/interface.rs, line 263 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAj8hrhrpjviTaSkxE-:-KAj8hriABPkZKYp_4P5:-2001625322) ([raw file](https://github.com/servo/servo/blob/857cf8773e1cf1474228930c76d7e7b2159ee930/components/script/dom/bindings/interface.rs#L263)):*
Please name that variable `value` like in `has_instance`.

---

*[components/script/dom/bindings/interface.rs, line 287 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAj8qSA02eEue3iEI9v:-KAj8qSA02eEue3iEI9w:1549124375) ([raw file](https://github.com/servo/servo/blob/857cf8773e1cf1474228930c76d7e7b2159ee930/components/script/dom/bindings/interface.rs#L287)):*
`interface_object.get()`

---

*[components/script/dom/bindings/interface.rs, line 288 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAj8nImnNeLRmmGZd8n:-KAj8nImnNeLRmmGZd8o:1953170574) ([raw file](https://github.com/servo/servo/blob/857cf8773e1cf1474228930c76d7e7b2159ee930/components/script/dom/bindings/interface.rs#L288)):*
Am stupid, `&js_class` should be just `js_class`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9633#-:-KAj8xM0TVAoD_BN08yk:251022119)*

**Comment by nox** [comment]
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[components/script/dom/bindings/interface.rs, line 290 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAjHvFPVLqYSmYVCypX:-KAjHvFPVLqYSmYVCypY:-403109264) ([raw file](https://github.com/servo/servo/blob/857cf8773e1cf1474228930c76d7e7b2159ee930/components/script/dom/bindings/interface.rs#L290)):*
Remove that.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9633)*

**Comment by nox** [comment]
-S-tests-failed +S-needs-code-changes 

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[components/script/dom/bindings/interface.rs, line 260 \[r5\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAnnC1hV82-5QPDrP-G:-KAnnC1hV82-5QPDrP-H:-844226326) ([raw file](https://github.com/servo/servo/blob/c14ba3551b3c8260bf4e568e28886a2c690793fd/components/script/dom/bindings/interface.rs#L260)):*
Nit: period at the end.

---

*[components/script/dom/bindings/interface.rs, line 287 \[r5\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAnmwgXtWN1q46VLj0f:-KAnmwgXtWN1q46VLj0g:-132570608) ([raw file](https://github.com/servo/servo/blob/c14ba3551b3c8260bf4e568e28886a2c690793fd/components/script/dom/bindings/interface.rs#L287)):*
My previous comment meant to replace `*interface_object.ptr` by `interface_object.get()`, not the whole expression. The current code makes little sense, doesn't it?

---

*[components/script/dom/bindings/interface.rs, line 288 \[r5\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAnn4Hvi0XAmXwku3y_:-KAnn4Hvi0XAmXwku3ya:1283786899) ([raw file](https://github.com/servo/servo/blob/c14ba3551b3c8260bf4e568e28886a2c690793fd/components/script/dom/bindings/interface.rs#L288)):*
Remove `as *const _`, JS_GetClass already returns a `*const JSClass`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9633#-:-KAnnFc18F8XLjxTu-Ho:383483901)*

**Comment by nox** [comment]
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[components/script/dom/bindings/interface.rs, line 288 \[r5\]](https://reviewable.io:443/reviews/servo/servo/9633#-KAnn4Hvi0XAmXwku3y_:-KApxltPQwt6HyeGSdJf:-333069921) ([raw file](https://github.com/servo/servo/blob/c14ba3551b3c8260bf4e568e28886a2c690793fd/components/script/dom/bindings/interface.rs#L288)):*
No no no, I meant this:

```rust
let js_class = JS_GetClass(interface_object.get());
let object_class = &*(js_class as *const NonCallbackInterfaceObjectClass);
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9633#-:-KApxx8yXaeJcqWbqIwo:1458266033)*

**Comment by nox** [comment]
@bors-servo r+

LGTM!

---

Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9633#-:-KApzXthtmPlANoZG38Q:-130662487)*

**Comment by KiChjang** [comment]
@bors-servo retry
* #8127

**Comment by KiChjang** [comment]
@bors-servo retry
* #9148

**Comment by KiChjang** [comment]
@bors-servo retry
* #8191

**Comment by KiChjang** [comment]
@bors-servo retry
* #9063

**Comment by KiChjang** [comment]
@bors-servo retry
* #8191, #8127

**Pull request merged**